src/fides/api/task/graph_task.py

```diff
@@ -585,56 +587,74 @@ def erasure_request(
+        if not self.execution_node.collection.contains_field(lambda f: f.primary_key):
             logger.warning(
                 "No erasures on {} as there is no primary_key defined.",
-                self.traversal_node.node.address,
+                self.execution_node.address,
             )
+            if self.request_task.id:
+                # DSR 3.0
+                self.request_task.rows_masked = 0  # Saved as part of update_status
+
+            # TODO Remove when we stop support for DSR 2.0
+            self.resources.cache_erasure(
+                f"{self.key}", 0
+            )  # Cache that the erasure was performed in case we need to restart
             self.update_status(
                 "No values were erased since no primary key was defined for this collection",
                 None,
                 ActionType.erasure,
                 ExecutionLogStatus.complete,
             )
-            # Cache that the erasure was performed in case we need to restart
-            self.resources.cache_erasure(self.key.value, 0)
             return 0
 
         if not self.can_write_data():
             logger.warning(
                 "No erasures on {} as its ConnectionConfig does not have write access.",
-                self.traversal_node.node.address,
+                self.execution_node.address,
             )
+            if self.request_task.id:
+                # DSR 3.0
+                self.request_task.rows_masked = 0  # Saved as part of update_status
+
+            # TODO Remove when we stop support for DSR 2.0
+            self.resources.cache_erasure(
+                f"{self.key}", 0
+            )  # Cache that the erasure was performed in case we need to restart
             self.update_status(
                 f"No values were erased since this connection {self.connector.configuration.key} has not been "
                 f"given write access",
                 None,
                 ActionType.erasure,
                 ExecutionLogStatus.error,
             )
-            self.resources.cache_erasure(self.key.value, 0)
             return 0
 
         formatted_input_data: NodeInput = self.pre_process_input_data(
             *inputs, group_dependent_fields=True
         )
 
         output = self.connector.mask_data(
-            self.traversal_node,
+            self.execution_node,
             self.resources.policy,
             self.resources.request,
+            self.resources.privacy_request_task,
             retrieved_data,
             formatted_input_data,
         )
-        self.log_end(ActionType.erasure)
+        if self.request_task.id:
+            # DSR 3.0
+            self.request_task.rows_masked = 0  # Saved as part of update_status
+        # TODO Remove when we stop support for DSR 2.0
         self.resources.cache_erasure(
             f"{self.key}", output
         )  # Cache that the erasure was performed in case we need to restart
+        self.log_end(ActionType.erasure)
```

Comment:
DSR 3.0 saving rows masked directly on the RequestTask while DSR 2.0 is caching it. This is largely used for testing, and for DSR 2.0 further kept us from re-running an erasure node that had already run.  DSR 3.0 has persistent tasks so can rely on the status of the task itself.